### ISSUE TITLE
Cannot compile android export templates with Mono in the master branch

### ISSUE THREAD

**Issue by Andrey722** [godotengine/godot#43936]
**Godot version:**
Master branch 3b4fad4
Also tried 3.2 stable before (more in description)

**OS/device including version:**
Windows 10 64-bit
VS 2019 Native tools Command Prompt x64

**Issue description:**
Firstly I compiled the stable branch of 3.2 with mono enabled (following the instructions)
Also compiled mono for android (https://github.com/godotengine/godot-mono-builds)
Then in VS 2019 Tools I'm compiling templates: set env variables, running command scons p=android target=release module_mono_enabled=yes android_arch=armv7 mono_prefix="Path-To-Compiled-Monos\android-armeabi-v7a-release"
And on file `Godot-Source-Path\modules\mono\mono_gd\android_mono_config.gen.cpp` I have this issue:
```
[ 18%] Compiling ==> modules\mono\mono_gd\android_mono_config.gen.cpp
=====
"modules\\mono\\mono_gd\\android_mono_config.gen.cpp:21:27: error: member reference base type 'uint8_t *' (aka 'unsigned char *') is not a structure or union
Compression::decompress(w.ptr(), config_uncompressed_size, config_compressed_data,
~^~~~
modules\\mono\\mono_gd\\android_mono_config.gen.cpp:24:34: error: member reference base type 'uint8_t *' (aka 'unsigned char *') is not a structure or union
if (s.parse_utf8((const char *)w.ptr(), data.size())) {
~^~~~
2 errors generated.
"
=====
scons: *** [modules\mono\mono_gd\android_mono_config.gen.android.opt.armv7.neon.o] Error 1
scons: building terminated because of errors.
```
I've started to search in source code and found the code, what generates the `android_mono_config.gen.cpp`.
It was generated by `Godot-Source-Path\modules\mono\build_scripts\make_android_mono_config.py`

Before I tried to compile stable 3.2 Godot branch. But there I had another problems.
Is was about 20 errors generated on compiling file `Godot-Source-Path\modules\mono\csharp_script.cpp`, but in master this file compiles good. And `android_mono_config.gen.cpp` generation in 3.2 was different with current master too.
So I hope its not so big problem and you will fix it fast or maybe help me if its problem in my compilation process..
I hope I wrote enough info... Thanks in advance :)

**Steps to reproduce:**
1. Compile 3b4fad4 Godot master
2. Compile Mono for Android from https://github.com/mono/mono (master branch)
3. Compile Android Templates with scons p=android target=release module_mono_enabled=yes android_arch=armv7 mono_prefix="Path-To-Compiled-Monos\android-armeabi-v7a-release"

**Comment by Calinou** [comment]
cc @neikeq

**Comment by Andrey722** [comment]
Any ideas?..

**Comment by akien-mga** [comment]
If I understand correctly you compiled first `3.2` then `master`? If so you should clean you build artifacts and generated files as they're not compatible. You can do it with `git clean -fxd` (removes all untracked files).

We advise to use different Git clones if you want to compile `3.2` and `master` in parallel.

**Comment by Andrey722** [comment]
@akien-mga, ofcourse I did it in different clones, but it didn't affect.

**Comment by akien-mga** [comment]
So you still have the same issue compiling `master` after `git clean -fxd`?

**Comment by Andrey722** [comment]
@akien-mga I just loaded .zip from site, so I cant use `git clean -fxd`. Do I need to clone it with git commands and make `git clean -fxd`?

**Comment by akien-mga** [comment]
No if it's from a ZIP, you can just re-extract the zip to have a clean build environment.

**Comment by Andrey722** [comment]
Ok, I will try to do it, but I think it won't be useful, because I compiled every time with just re-extracted zip. Maybe I missed something and it will help. I need about 2 hours I think..

**Comment by akien-mga** [comment]
I'll see if I can reproduce it too.

**Comment by akien-mga** [comment]
OK I could reproduce the issue, so no need for you to try again. My assumption was wrong in https://github.com/godotengine/godot/issues/43936#issuecomment-737785907, it's not a user-end issue, it's the Mono Android config being incompatible with recent `master` changes to core APIs.

**Comment by Andrey722** [comment]
Ok, thanks. I'll wait a fixing patch ;)

**Comment by akien-mga** [comment]
#44064 should fix it.

Note that the `master` branch is for Godot 4.0 which is still a long way away from a stable release, and both Mono and Android are not fully up-to-date to recent core changes. So if you're interested in testing and debugging, you're welcome to try them out, but if you're looking into compiling Godot with Mono for an actual game project, I would strongly recommend using the `3.2` branch for now.

**Comment by Andrey722** [comment]
As I said 3.2 branch crushes build before it reaches `android_mono_config.gen.cpp`. It crushes on `modules\mono\csharp_script.cpp`. If you'll wait I can provide full errors list (about 20 errors generated)

**Comment by akien-mga** [comment]
Please do so in a different issue, this one is specific to a `master`-only issue that is now fixed.

**Comment by Andrey722** [comment]
Ok, finally) I loaded latest 3.2 and compiled it, then tried to compile templates:
```
[ 17%] Compiling ==> modules\mono\csharp_script.cpp
=====
"In file included from modules\\mono\\csharp_script.cpp:31:
In file included from modules\\mono/csharp_script.h:39:
modules\\mono/mono_gc_handle.h:38:7: error: definition of type 'MonoGCHandle' conflicts with typedef of the same name
class MonoGCHandle : public Reference {
C:\\Program Files\\Mono\\Mono-installs\\android-armeabi-v7a-release\\include\\mono-2.0\\mono/utils/mono-forward.h:13:16: note: 'MonoGCHandle' declared here
typedef void * MonoGCHandle;
               ^
In file included from modules\\mono\\csharp_script.cpp:58:
modules\\mono/signal_awaiter_utils.h:42:36: error: base specifier must name a class
class SignalAwaiterHandle : public MonoGCHandle {
                            ~~~~~~~^~~~~~~~~~~~
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
GDCLASS(SignalAwaiterHandle, MonoGCHandle);
                                     ^
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
modules\\mono/signal_awaiter_utils.h:44:2: error: use of undeclared identifier '_get_category'
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
        ^
.\\core/object.h:295:7: note: expanded from macro 'GDCLASS'
                if (_get_category != m_inherits::_get_category) {                                                                               \\
                    ^
In file included from modules\\mono\\csharp_script.cpp:58:
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
                                     ^
modules\\mono/signal_awaiter_utils.h:44:2: error: use of undeclared identifier '_get_category'
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
        ^
.\\core/object.h:298:16: note: expanded from macro 'GDCLASS'
                        category += _get_category();                                                                                                \\
                                    ^
In file included from modules\\mono\\csharp_script.cpp:58:
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
                                     ^
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
modules\\mono/signal_awaiter_utils.h:44:2: error: no member named '_get_valid_parents_static' in 'SignalAwaiterHandle'; did you mean 'get_valid_parents_static'?
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\\core/object.h:310:16: note: expanded from macro 'GDCLASS'
                if (m_class::_get_valid_parents_static != m_inherits::_get_valid_parents_static) {                                              \\
                    ~~~~~~~~~^
modules\\mono/signal_awaiter_utils.h:44:2: note: 'get_valid_parents_static' declared here
.\\core/object.h:308:14: note: expanded from macro 'GDCLASS'
        static void get_valid_parents_static(List<String> *p_parents) {                                                                     \\
                    ^
In file included from modules\\mono\\csharp_script.cpp:58:
modules\\mono/signal_awaiter_utils.h:44:2: error: no type named '_get_valid_parents_static' in 'SignalAwaiterHandle'
        GDCLASS(SignalAwaiterHandle, MonoGCHandle)
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\\core/object.h:311:13: note: expanded from macro 'GDCLASS'
                        m_class::_get_valid_parents_static(p_parents);                                                                              \\
                        ~~~~~~~~~^
In file included from modules\\mono\\csharp_script.cpp:58:
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
                                     ^
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
modules\\mono/signal_awaiter_utils.h:44:10: error: no member named '_get' in 'SignalAwaiterHandle'; did you mean '::SignalAwaiterHandle::_getv'?
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);        ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\\core/object.h:339:68: note: expanded from macro 'GDCLASS'
                return (bool (Object::*)(const StringName &, Variant &) const) & m_class::_get;                                                 \\
                                                                                 ^~~~~~~~~
modules\\mono/signal_awaiter_utils.h:44:2: note: '::SignalAwaiterHandle::_getv' declared here
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
        ^
.\\core/object.h:341:15: note: expanded from macro 'GDCLASS'
        virtual bool _getv(const StringName &p_name, Variant &r_ret) const {                                                                \\
                     ^
In file included from modules\\mono\\csharp_script.cpp:58:
modules\\mono/signal_awaiter_utils.h:44:31: error: 'MonoGCHandle' (aka 'void *') is not a class, namespace, or enumeration
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
                                     ^
modules\\mono/signal_awaiter_utils.h:44:2: error: use of undeclared identifier '_get'; did you mean '_getv'?
        GDCLASS(SignalAwaiterHandle, MonoGCHandle);
        ^
.\\core/object.h:343:8: note: expanded from macro 'GDCLASS'
                        if (_get(p_name, r_ret))                                                                                                    \\
                            ^
modules\\mono/signal_awaiter_utils.h:44:2: note: '_getv' declared here
.\\core/object.h:341:15: note: expanded from macro 'GDCLASS'
        virtual bool _getv(const StringName &p_name, Variant &r_ret) const {                                                                \\
                     ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]\r\n20 errors generated.
"
=====
scons: *** [modules\mono\csharp_script.android.opt.armv7.neon.o] Error 1
scons: building terminated because of errors.
```

**Comment by akien-mga** [comment]
Please open a new issue as advised, while it might seem to be the same from an end user point of view (can't build templates), the errors are unrelated and the bugs are separate. The one tracked in this issue affected the `master` branch only and is now fixed.

**Comment by Andrey722** [comment]
Oh sorry, I skipped about new issue.. I'll open it
